src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,539 @@
+            {
+                throw new ArgumentException("MONGODB-AWS authentication may only use the $external source.", nameof(credential));
+            }
+
+            return CreateMechanism(credential.Username, credential.Password, properties, randomByteGenerator, clock);
+        }
+
+        private static MongoAWSMechanism CreateMechanism(
+            string username,
+            SecureString securePassword,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock clock)
+        {
+            ValidateMechanismProperties(properties);
+
+            var awsCredentials =
+                CreateAwsCredentialsFromMongoCredentials(username, securePassword, properties) ??
+                CreateAwsCredentialsFromEnvironmentVariables() ??
+                CreateAwsCredentialsFromEcsResponse() ??
+                CreateAwsCredentialsFromEc2Response();
+
+            if (awsCredentials == null)
+            {
+                throw new InvalidOperationException("A MONGODB-AWS credential must have an access key ID.");
+            }
+
+            var credentials = new UsernamePasswordCredential("$external", awsCredentials.Username, awsCredentials.Password);
+
+            return new MongoAWSMechanism(credentials, awsCredentials.SessionToken, randomByteGenerator, clock);
+        }
+
+        private static AwsCredentials CreateAwsCredentialsFromMongoCredentials(string username, SecureString securePassword, IEnumerable<KeyValuePair<string, string>> properties)
+        {
+            var sessionToken = ExtractSessionTokenFromMechanismProperties(properties);
+
+            var awsCredentials = new AwsCredentials
+            {
+                Username = username,
+                Password = securePassword,
+                SessionToken = sessionToken
+            };
+
+            ValidateAwsCredentials(awsCredentials);
+
+            if (awsCredentials.Username == null)
+            {
+                return null;
+            }
+            
+            return awsCredentials;
+        }
+
+        private static AwsCredentials CreateAwsCredentialsFromEnvironmentVariables()
+        {
+            var username = Environment.GetEnvironmentVariable("AWS_ACCESS_KEY_ID");
+            var password = Environment.GetEnvironmentVariable("AWS_SECRET_ACCESS_KEY");
+            var sessionToken = Environment.GetEnvironmentVariable("AWS_SESSION_TOKEN");
+
+            var awsCredentials = new AwsCredentials
+            {
+                Username = username,
+                Password = ToSecureString(password),
+                SessionToken = sessionToken
+            };
+
+            ValidateAwsCredentials(awsCredentials);
+
+            if (awsCredentials.Username == null)
+            {
+                return null;
+            }
+
+            return awsCredentials;
+        }
+
+        private static AwsCredentials CreateAwsCredentialsFromEcsResponse()
```

Comment:
Refactor this method to:

    private static AwsCredentials CreateAwsCredentialsFromEcsResponse()
    {
        var relativeUri = Environment.GetEnvironmentVariable("AWS_CONTAINER_CREDENTIALS_RELATIVE_URI");
        if (relativeUri == null)
        {
            return null;
        }

        var response = AwsHttpClientHelper.GetECSResponseAsync(relativeUri).GetAwaiter().GetResult();
        var parsedResponse = BsonDocument.Parse(response);
        var accessKeyId = parsedResponse.GetValue("AccessKeyId", null)?.AsString;
        var secretAccessKey = parsedResponse.GetValue("SecretAccessKey", null)?.AsString;
        var sessionToken = parsedResponse.GetValue("Token", null)?.AsString;

        return new AwsCredentials(accessKeyId, ToSecureString(secretAccessKey), sessionToken);
    }